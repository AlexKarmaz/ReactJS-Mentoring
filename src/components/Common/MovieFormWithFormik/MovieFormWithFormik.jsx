import React, { useCallback } from 'react';
import StyledButton from '../StyledButton';
import LabeledInput from '../LabeledInput';
import LabeledMultiSelect from '../LabeledMultiSelect';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import './MovieFormWithFormik.css';

const MovieFormWithFormik = ({ onSubmit, movieData = {}, genres }) => {
    const onFormSubmit = useCallback((data) => onSubmit(data), [onSubmit]);

    const formik = useFormik({
        initialValues: movieData,
        validationSchema: Yup.object({
          title: Yup.string().required('Title is required'),
          release_date: Yup.string()
            .required('Release date is required'),
          poster_path: Yup.string().url('Poster should be a valid url').required('Poster is required'),
          genres: Yup.array().min(1).required('At least 1 genre must be selected'),
          overview: Yup.string().required('Movie description is required'),
          runtime: Yup.number()
            .min(0, 'Runtime should be more then 0' )
            .required('Runtime is required'),
        }),
        onSubmit: onFormSubmit,
    });

    const modifiedGenres = genres.map((genre) => ({label: genre, value: genre}));

    const onGenreChange = useCallback ((items) => {
        const genres = items.map((item) => item.value);
        formik.setFieldValue('genres', genres, true);
    }, [formik]);

    const getErrorMessageForField = useCallback(
        (fieldName) => {
          if (formik.touched[fieldName] && formik.errors[fieldName]) {
            return formik.errors[fieldName];
          }
          return '';
        },
        [formik],
    );

    return (
        <form className='movieForm' onSubmit={formik.handleSubmit}>
            {formik.values.id && (<LabeledInput
                name='id'
                title='Movie id'
                value={formik.values.id}
                readOnly={true}
            />)}
            <LabeledInput
                name='title'
                title='Title'
                value={formik.values.title}
                onChange={formik.handleChange}
                placeholder='Title here'
                onBlur={formik.handleBlur}
                validationError={getErrorMessageForField('title')}
            />
            <LabeledInput
                name='release_date'
                title='Release date'
                type='date'
                value={formik.values.release_date}
                onChange={formik.handleChange}
                onBlur={formik.handleBlur}
                validationError={getErrorMessageForField('release_date')}
            />
            <LabeledInput
                name='poster_path'
                title='Movie url'
                type='url'
                value={formik.values.poster_path}
                onChange={formik.handleChange}
                placeholder='Movie URL here'
                onBlur={formik.handleBlur}
                validationError={getErrorMessageForField('poster_path')}
            />
            <LabeledMultiSelect
                title="Genres"
                options={modifiedGenres}
                onChange={onGenreChange}
                selected={formik.values.genres}
                error={getErrorMessageForField('genres')}
                overrideStrings={
                    {'selectSomeItems': 'Select Genre',}
                }
            />
            <LabeledInput
                name='overview'
                title='Overview'
                value={formik.values.overview}
                onChange={formik.handleChange}
                placeholder='Overview here'
                onBlur={formik.handleBlur}
                validationError={getErrorMessageForField('overview')}
            />
            <LabeledInput
                name='runtime'
                title='Runtime'
                type='number'
                value={formik.values.runtime}
                onChange={formik.handleChange}
                placeholder='Runtime here'
                onBlur={formik.handleBlur}
                validationError={getErrorMessageForField('runtime')}
            />
            <div className='movieForm-footer'>
                <StyledButton
                    text='Reset'
                    size='medium'
                    type='reset'
                    onClick={formik.resetForm}
                />
                <StyledButton text='Submit' size='medium' type='confirm' onClick={formik.handleSubmit} />
            </div>
        </form>
    );
};

export default MovieFormWithFormik;
