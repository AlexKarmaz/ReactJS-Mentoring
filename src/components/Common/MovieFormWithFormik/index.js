import MovieFormWithFormik from './MovieFormWithFormik';

export default MovieFormWithFormik;
