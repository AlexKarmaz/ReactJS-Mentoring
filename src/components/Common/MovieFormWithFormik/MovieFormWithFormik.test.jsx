import React from 'react';
import { render, fireEvent, waitFor, cleanup, screen } from '@testing-library/react';
import MovieFormWithFormik from './MovieFormWithFormik';

describe('MovieFormWithFormik', () => {
    beforeEach(async () => {
        jest.resetAllMocks();
    });

    it('should be rendered correctly', () => {
        const onSubmit = jest.fn();
        const form = render(
            <MovieFormWithFormik
                onSubmit={onSubmit}
                genres={['Comedy', 'Horror']}
            />
        );

        expect(form).toMatchSnapshot();
    });

    it('onSubmit should not be called if fields are empty', async () => {
        const onSubmit = jest.fn();

        const form = render(
            <MovieFormWithFormik
                onSubmit={onSubmit}
                genres={['Comedy', 'Horror']}
            />
        );

        fireEvent.click(screen.getByText('Submit'));
        await waitFor(() => {
            expect(onSubmit).not.toHaveBeenCalled();
        });
    });

    it('errors should be displayed if all fields are empty and onSubmit is called', async () => {
        const onSubmit = jest.fn();

        const form = render(
            <MovieFormWithFormik
                onSubmit={onSubmit}
                genres={['Comedy', 'Horror']}
            />
        );

        expect(
            form.container.querySelectorAll('p.labeledInput-error').length
        ).toEqual(0);

        fireEvent.blur(screen.getByPlaceholderText('Title here'));
        fireEvent.blur(screen.getByPlaceholderText('Movie URL here'));
        fireEvent.blur(screen.getByPlaceholderText('Overview here'));

        await waitFor(() => {
            expect(
                form.container.querySelectorAll('p.labeledInput-error').length
            ).toEqual(3);
        });

        expect(form).toMatchSnapshot();
    });

    it('onSubmit should be called if all fields are populated', async () => {
        const onSubmit = jest.fn();
        const initialState = {
            id:'1',
            title:'test',
            release_date:'2020-10-30',
            poster_path:'https://test.com',
            genres:['Comedy'],
            overview:'test',
            runtime:100,
        };
        const form = render(
            <MovieFormWithFormik
                movieData={initialState}
                onSubmit={onSubmit}
                genres={['Comedy', 'Horror']}
            />
        );

        fireEvent.click(screen.getByText('Submit'));

        await waitFor(() => {
            expect(
                form.container.querySelectorAll('p.labeled-input__error').length
            ).toEqual(0);
        });
        expect(
            form.container.querySelectorAll('.genres__error').length
        ).toEqual(0);

        expect(onSubmit).toHaveBeenCalled();
        expect(form).toMatchSnapshot();
    });
});
