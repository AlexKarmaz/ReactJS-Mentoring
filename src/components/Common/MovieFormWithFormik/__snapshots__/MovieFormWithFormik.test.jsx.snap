// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieFormWithFormik errors should be displayed if all fields are empty and onSubmit is called 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="movieForm"
      >
        <label>
          Title
          <input
            name="title"
            placeholder="Title here"
            type="text"
            value=""
          />
        </label>
        <p
          class="labeledInput-error"
        >
          Title is required
        </p>
        <label>
          Release date
          <input
            name="release_date"
            type="date"
            value=""
          />
        </label>
        <label>
          Movie url
          <input
            name="poster_path"
            placeholder="Movie URL here"
            type="url"
            value=""
          />
        </label>
        <p
          class="labeledInput-error"
        >
          Poster is required
        </p>
        <label>
          Genres
          <div
            class="go2139093995 multi-select"
          >
            <div
              aria-expanded="false"
              aria-readonly="true"
              class="go2646822163 dropdown-container"
              tabindex="0"
            >
              <div
                class="go2642161244 dropdown-heading"
              >
                <div
                  class="dropdown-heading-value"
                >
                  <span
                    class="gray"
                  >
                    Select Genre
                  </span>
                </div>
                <svg
                  class="dropdown-heading-dropdown-arrow gray"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-width="2"
                  width="24"
                >
                  <path
                    d="M6 9L12 15 18 9"
                  />
                </svg>
              </div>
            </div>
          </div>
        </label>
        <label>
          Overview
          <input
            name="overview"
            placeholder="Overview here"
            type="text"
            value=""
          />
        </label>
        <p
          class="labeledInput-error"
        >
          Movie description is required
        </p>
        <label>
          Runtime
          <input
            name="runtime"
            placeholder="Runtime here"
            type="number"
            value=""
          />
        </label>
        <div
          class="movieForm-footer"
        >
          <button
            class="styledButton undefined medium reset"
            type="button"
          >
            Reset
          </button>
          <button
            class="styledButton undefined medium confirm"
            type="button"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="movieForm"
    >
      <label>
        Title
        <input
          name="title"
          placeholder="Title here"
          type="text"
          value=""
        />
      </label>
      <p
        class="labeledInput-error"
      >
        Title is required
      </p>
      <label>
        Release date
        <input
          name="release_date"
          type="date"
          value=""
        />
      </label>
      <label>
        Movie url
        <input
          name="poster_path"
          placeholder="Movie URL here"
          type="url"
          value=""
        />
      </label>
      <p
        class="labeledInput-error"
      >
        Poster is required
      </p>
      <label>
        Genres
        <div
          class="go2139093995 multi-select"
        >
          <div
            aria-expanded="false"
            aria-readonly="true"
            class="go2646822163 dropdown-container"
            tabindex="0"
          >
            <div
              class="go2642161244 dropdown-heading"
            >
              <div
                class="dropdown-heading-value"
              >
                <span
                  class="gray"
                >
                  Select Genre
                </span>
              </div>
              <svg
                class="dropdown-heading-dropdown-arrow gray"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-width="2"
                width="24"
              >
                <path
                  d="M6 9L12 15 18 9"
                />
              </svg>
            </div>
          </div>
        </div>
      </label>
      <label>
        Overview
        <input
          name="overview"
          placeholder="Overview here"
          type="text"
          value=""
        />
      </label>
      <p
        class="labeledInput-error"
      >
        Movie description is required
      </p>
      <label>
        Runtime
        <input
          name="runtime"
          placeholder="Runtime here"
          type="number"
          value=""
        />
      </label>
      <div
        class="movieForm-footer"
      >
        <button
          class="styledButton undefined medium reset"
          type="button"
        >
          Reset
        </button>
        <button
          class="styledButton undefined medium confirm"
          type="button"
        >
          Submit
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MovieFormWithFormik onSubmit should be called if all fields are populated 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="movieForm"
      >
        <label>
          Movie id
          <input
            name="id"
            readonly=""
            type="text"
            value="1"
          />
        </label>
        <label>
          Title
          <input
            name="title"
            placeholder="Title here"
            type="text"
            value="test"
          />
        </label>
        <label>
          Release date
          <input
            name="release_date"
            type="date"
            value="2020-10-30"
          />
        </label>
        <label>
          Movie url
          <input
            name="poster_path"
            placeholder="Movie URL here"
            type="url"
            value="https://test.com"
          />
        </label>
        <label>
          Genres
          <div
            class="go2139093995 multi-select"
          >
            <div
              aria-expanded="false"
              aria-readonly="true"
              class="go2646822163 dropdown-container"
              tabindex="0"
            >
              <div
                class="go2642161244 dropdown-heading"
              >
                <div
                  class="dropdown-heading-value"
                >
                  <span>
                    Comedy
                  </span>
                </div>
                <svg
                  class="dropdown-heading-dropdown-arrow gray"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-width="2"
                  width="24"
                >
                  <path
                    d="M6 9L12 15 18 9"
                  />
                </svg>
              </div>
            </div>
          </div>
        </label>
        <label>
          Overview
          <input
            name="overview"
            placeholder="Overview here"
            type="text"
            value="test"
          />
        </label>
        <label>
          Runtime
          <input
            name="runtime"
            placeholder="Runtime here"
            type="number"
            value="100"
          />
        </label>
        <div
          class="movieForm-footer"
        >
          <button
            class="styledButton undefined medium reset"
            type="button"
          >
            Reset
          </button>
          <button
            class="styledButton undefined medium confirm"
            type="button"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="movieForm"
    >
      <label>
        Movie id
        <input
          name="id"
          readonly=""
          type="text"
          value="1"
        />
      </label>
      <label>
        Title
        <input
          name="title"
          placeholder="Title here"
          type="text"
          value="test"
        />
      </label>
      <label>
        Release date
        <input
          name="release_date"
          type="date"
          value="2020-10-30"
        />
      </label>
      <label>
        Movie url
        <input
          name="poster_path"
          placeholder="Movie URL here"
          type="url"
          value="https://test.com"
        />
      </label>
      <label>
        Genres
        <div
          class="go2139093995 multi-select"
        >
          <div
            aria-expanded="false"
            aria-readonly="true"
            class="go2646822163 dropdown-container"
            tabindex="0"
          >
            <div
              class="go2642161244 dropdown-heading"
            >
              <div
                class="dropdown-heading-value"
              >
                <span>
                  Comedy
                </span>
              </div>
              <svg
                class="dropdown-heading-dropdown-arrow gray"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-width="2"
                width="24"
              >
                <path
                  d="M6 9L12 15 18 9"
                />
              </svg>
            </div>
          </div>
        </div>
      </label>
      <label>
        Overview
        <input
          name="overview"
          placeholder="Overview here"
          type="text"
          value="test"
        />
      </label>
      <label>
        Runtime
        <input
          name="runtime"
          placeholder="Runtime here"
          type="number"
          value="100"
        />
      </label>
      <div
        class="movieForm-footer"
      >
        <button
          class="styledButton undefined medium reset"
          type="button"
        >
          Reset
        </button>
        <button
          class="styledButton undefined medium confirm"
          type="button"
        >
          Submit
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`MovieFormWithFormik should be rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="movieForm"
      >
        <label>
          Title
          <input
            name="title"
            placeholder="Title here"
            type="text"
            value=""
          />
        </label>
        <label>
          Release date
          <input
            name="release_date"
            type="date"
            value=""
          />
        </label>
        <label>
          Movie url
          <input
            name="poster_path"
            placeholder="Movie URL here"
            type="url"
            value=""
          />
        </label>
        <label>
          Genres
          <div
            class="go2139093995 multi-select"
          >
            <div
              aria-expanded="false"
              aria-readonly="true"
              class="go2646822163 dropdown-container"
              tabindex="0"
            >
              <div
                class="go2642161244 dropdown-heading"
              >
                <div
                  class="dropdown-heading-value"
                >
                  <span
                    class="gray"
                  >
                    Select Genre
                  </span>
                </div>
                <svg
                  class="dropdown-heading-dropdown-arrow gray"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-width="2"
                  width="24"
                >
                  <path
                    d="M6 9L12 15 18 9"
                  />
                </svg>
              </div>
            </div>
          </div>
        </label>
        <label>
          Overview
          <input
            name="overview"
            placeholder="Overview here"
            type="text"
            value=""
          />
        </label>
        <label>
          Runtime
          <input
            name="runtime"
            placeholder="Runtime here"
            type="number"
            value=""
          />
        </label>
        <div
          class="movieForm-footer"
        >
          <button
            class="styledButton undefined medium reset"
            type="button"
          >
            Reset
          </button>
          <button
            class="styledButton undefined medium confirm"
            type="button"
          >
            Submit
          </button>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="movieForm"
    >
      <label>
        Title
        <input
          name="title"
          placeholder="Title here"
          type="text"
          value=""
        />
      </label>
      <label>
        Release date
        <input
          name="release_date"
          type="date"
          value=""
        />
      </label>
      <label>
        Movie url
        <input
          name="poster_path"
          placeholder="Movie URL here"
          type="url"
          value=""
        />
      </label>
      <label>
        Genres
        <div
          class="go2139093995 multi-select"
        >
          <div
            aria-expanded="false"
            aria-readonly="true"
            class="go2646822163 dropdown-container"
            tabindex="0"
          >
            <div
              class="go2642161244 dropdown-heading"
            >
              <div
                class="dropdown-heading-value"
              >
                <span
                  class="gray"
                >
                  Select Genre
                </span>
              </div>
              <svg
                class="dropdown-heading-dropdown-arrow gray"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-width="2"
                width="24"
              >
                <path
                  d="M6 9L12 15 18 9"
                />
              </svg>
            </div>
          </div>
        </div>
      </label>
      <label>
        Overview
        <input
          name="overview"
          placeholder="Overview here"
          type="text"
          value=""
        />
      </label>
      <label>
        Runtime
        <input
          name="runtime"
          placeholder="Runtime here"
          type="number"
          value=""
        />
      </label>
      <div
        class="movieForm-footer"
      >
        <button
          class="styledButton undefined medium reset"
          type="button"
        >
          Reset
        </button>
        <button
          class="styledButton undefined medium confirm"
          type="button"
        >
          Submit
        </button>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
